keps/sig-network/1111-dynamically-re-configure-lb-node-set/README.md

```diff
@@ -0,0 +1,681 @@
+looking at the generic cloud-provider API for load balancer operations that the
+KCCM can invoke: the only functions which ends up getting called when a node is
+deleted is: `EnsureLoadBalancer`/`UpdateLoadBalancer`. But the list of nodes
+provided as an argument to both (`nodes`) should contain __all the remaining
+ cluster nodes__. It is therefor impossible that deleting a Node from the LB set
+__after__ the VM has been deleted, should cause any problem. The cloud-provider
+API can only be invoked using all the remaining and known Nodes/VMs. Either the
+reasons for adding this condition in the first place was misunderstood, or they
+were wrong or they might have been right __at the time__, but with the current
+code it certainly doesn't make sense to keep.
+
+The second predicate has no direct correlation to networking, but __could__,
+__possibly__, __maybe__ indicate a general problem with the node should the
+Kubelet be un-healthy in certain specific cases (for ex: if it's resource
+starved, which could be impacting other system processes as well, and
+subsequently networking related processes).
+
+We can define two classes of services, when the type is `LoadBalancer`.
+
+- A) services which have `externalTrafficPolicy: Cluster`
+- B) services which have `externalTrafficPolicy: Local`
+
+These are the only two classes of services which impact traffic forwarding for a
+load balancer and answer the question: "where traffic should be sent to?".
+
+Load balancers, today, go about figuring out where "traffic should be sent to"
+through the utilization of an HC. The HC is configured differently depending on
+the class of service.
+
+- For services of class B: the HC is configured to probe all nodes in its set
+  on a node port defined to be value of the field `healthCheckNodePort`. The
+  field can be user defined, but when its not: it's automatically generated by
+  Kubernetes. The entity answering to the HC on this port is the service proxy
+  and its answer currently reflects one sole criteria: does the service have an
+  endpoint running on the node in question? If the answer is "yes", then an HTTP
+  status 200 (StatusOK) is returned to the LB, indicating to the LB that it can
+  use the node as a target for traffic forwarding. If the answer is "no", then
+  an HTTP status 503 (StatusServiceUnavailable) is returned, indicating the
+  inverse. Service proxies which implement and support class B type services,
+  must implement the above.
+- For services of class A: it's more of a "free for all". Cloud providers define
+  their own HC, and no one does the same. The following describes what the HC
+  looks like on the three major public cloud providers:
+  - GCP: probes port 10256 (Kube-proxy's healthz port)
+  - AWS: if ELB; probes the first `NodePort` defined on the service spec
+  - Azure: probes all `NodePort` defined on the service spec.
+  
+  All clouds take an approach of trying to ascertain if traffic can be forwarded
```

Comment:
yeah, it was more a thought in the sense, "if you have only one approach to do this everything will be simpler" ... but I can see that this may depende a lot on the specific implementations,